_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,257 @@
+Instead, the UTO does not store topic metadata and it aims to be scalable in terms of the number of topics that it can operate on.
+
+When running Kafka operations, the UTO makes use of the request batching supported by the Kafka Admin client to get higher throughput for metadata operations.
+All `KafkaTopic` events are queued when received, and then processed in batches by a number of controller threads (currently, only a single thread is supported).
+If more than one event related to a single topic resource are found in the batch building, they are put back in the queue to be processed with the next batch.
+
+The following line graph confirms the BTO scalability issue, while the UTO can scale almost linearly.
+Using the UTO with batch size 500 and linger 1000, we were able to achieve a max reconciliation time of 5.3 seconds with 10k concurrent topic events.
+
+<figure>
+    <img src="/assets/images/posts/2023-10-23-uto-max-recon-time.png" height=350>
+    <figcaption><small>
+        Fig 2. Line graph comparing BTO and UTO max reconciliation times.<br/>
+        We used a 3-nodes cluster with default configurations, running on a local Minikube instance with 10 cores and 28 GB of memory.<br/>
+        The test driver application was running on the same machine, and the time spent on the queue was not included.
+    </small></figcaption>
+</figure>
+
+You can tune the batching mechanism by setting `STRIMZI_MAX_QUEUE_SIZE` (default: 1024), `STRIMZI_MAX_QUEUE_SIZE` (default: 100), and `MAX_BATCH_LINGER_MS` (default: 100).
+If you exceed the configured max queue size, the UTO will print an error and then shutdown (Kubernetes will take care of restarting the pod).
+In that case, you can simply raise the max queues size to avoid the periodic operator restarts.
+
+### Race condition
+
+There is a known race condition happening when the operator and external applications try to change a topic configuration concurrently.
+Given the way Kafka is implemented, there is no solution to that problem, but the UTO significantly reduces the likelihood of reconciliation failures.
+
+When the external application wins, the topics will be created using the default Kafka cluster configuration.
+In this case, the BTO simply fails the reconciliation, and you need to recover manually.
+Instead, the UTO always reconciles, which results in reconfiguration only when the default configuration differs from the spec.
+That reconfiguration will either succeed, or fail.
+
+The recommendation here is to configure the Kafka cluster with `auto.create.topics.enable: false` in order to reduce the likelihood of hitting this problem.
+Unfortunately, disabling topic creation doesn't help with applications that use the Kafka Admin client to directly create required topics (i.e. Kafka Streams based applications).
+In this case, the application should be written to wait for topic existence, or use an init container to do the same.
```

Comment:
This is not my experience. Do you have any numbers to back this up?

From my experience, the BTO often creates/updates the `KafkaTopic` with the values from Kafka and is happy with it. Only the values are usually different from what the user requested.

I would argue the TO does not improve this much and the conflict will be still there. If the topic is auto-created with default settings, the KafkaTopic might change it and lead to keys being mixed up in partitions etc.

So I wonder if we should instead aim at disabling auto-creation in the first place without suggesting UTO helps here.

----

Also, is Kafka Streams really an issue here? It creates its own topic and users typically don't create them with KafkaTopic resources. So I don't think I ever herd about conflict for a Streams topic.

Reply:
Right, there is no difference between UTO and BTO with this issue.

A Streams application uses the admin client to create all required topics directly (CreateTopics API), and does not rely on auto topic creation.